### ISSUE TITLE
Add Entity generator 

### ISSUE THREAD

**Issue by sendilkumarn** [jhipster/jhipster-kotlin#45]

**Comment by sendilkumarn** [comment]
Yes 👍

**Comment by pvliss** [comment]
@sendilkumarn I was about to start working on this one but then I found there is already some work done as per issue #36 residing at branch [kotlin-entity](https://github.com/jhipster/jhipster-kotlin/tree/kotlin-entity). Does it mean it has been fixed already? If not, should I use what is already in that branch or start from scratch?

**Comment by sendilkumarn** [comment]
@pvliss From scratch and taking that branch as a reference will help. (if the branch is not in a mergeable state)

**Comment by pvliss** [comment]
@sendilkumarn  OK. I have already done some work through the weekend and today and hopefully will find some time to complete it tomorrow and create a new PR.

**Comment by sendilkumarn** [comment]
@pvliss Great go for it 👍

**Comment by pvliss** [comment]
Hello @sendilkumarn,

Sorry for the long delay but it needed way more work that I anticipated... :)

Anyways. I have done a first [commit](https://github.com/pvliss/jhipster-kotlin/commit/50b7f5c39b0c4aaf56d2b35298aedc54bf7ab86a) that you can check out.

I have been testing using the `sqlfull` sample from the parent JHipster project and have almost everything right, except the following issues:
1. There are still a few unit failing
2. Have not tested with gradle
3. Have not tested with the other samples.
4. We need a [PR](fix issue with relationship.otherEntityRelationshipName and entity generation #8992) accepted in the parent JHipster project in order for everything to work properly. 
5. We need to override the settings of the `maven-compiler-plugin` to disable annotation processing which should be **only** performed by kapt. I have tried using a hack(deleting the default and adding a new configuration) but does not always work and I d not like the approach. I think it would be best to handle from JHipster, e.g. add an extra variable e.g. `server_lang` and only generate the default when `server_lang==='java'` or otherwise add some kind of needle. What do you think?
6. I tried to use the same trick in `entity-server/files.js` as the one you used in the `server/files.js` to have all java or other files not defined as kt to be always copied but I am doing something wrong. All java files are always copied. Can you please also check what I am doing wrong. For that reason I have also copied the scala file for the Gatling test although I think this should come from the base generator instead to avoid maintenance.
7. Finally, regarding testing and CI I also wanted to ask if we are going to copy the samples and `*.json` entity files from the parent project to this repo and setup Travis to execute them all or can we re-use them in some way?

Regards,

**Comment by sendilkumarn** [comment]
Wow, that is great 👍 Yeah indeed it is a huge work. Create a PR we can do it together 
 
> 5. , e.g. add an extra variable e.g. server_lang and only generate the default when server_lang==='java' or otherwise add some kind of needle. What do you think?

I would say go for a needle. Create an issue in JHipster and let us discuss that there. Needle is easy.

> 6. Can you please also check what I am doing wrong.

Sure I will check it, once I have some time 👍 

> 7. Finally, regarding testing and CI I also wanted to ask if we are going to copy the samples and *.json entity files from the parent project to this repo and set up Travis to execute them all or can we re-use them in some way?

Sure this is always in my list of things to do. We can do that.

**Comment by pvliss** [comment]
> Create a PR we can do it together

You want me to create a PR even though it is unfinished?

> > 5. , e.g. add an extra variable e.g. server_lang and only generate the default when server_lang==='java' or otherwise add some kind of needle. What do you think?
> 
> I would say go for a needle. Create an issue in JHipster and let us discuss that there. Needle is easy.
 
Not sure if a needle will work though. From what I understand needles are mainly used to append and not to remove, but in this case we will need the following:
1. Remove the `<annotationProcessorPaths>` from the `<configuration>` element of the `maven-compiler-plugin`.
2. Add `<proc>none</proc>` to the the `<configuration>' element of the `maven-compiler-plugin`.
3. Finally, we need to make sure that the `kotlin-maven-plugin` takes hold of the process and according to the [kotlin maven docs](https://kotlinlang.org/docs/reference/using-maven.html#compiling-kotlin-and-java-sources)  we need to move the `maven-compiler-plugin` **after** the `kotlin-maven-plugin`.

So to summarize, the final `maven-compiler-plugin` configuration should be the following and declared **after** the `kotlin-maven-plugin`:
```xml
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <configuration>
                    <proc>none</proc>
                </configuration>
                <executions>
                    <!-- Replacing default-compile as it is treated specially by maven -->
                    <execution>
                        <id>default-compile</id>
                        <phase>none</phase>
                    </execution>
                    <!-- Replacing default-testCompile as it is treated specially by maven -->
                    <execution>
                        <id>default-testCompile</id>
                        <phase>none</phase>
                    </execution>
                    <execution>
                        <id>java-compile</id>
                        <phase>compile</phase>
                        <goals> <goal>compile</goal> </goals>
                    </execution>
                    <execution>
                        <id>java-test-compile</id>
                        <phase>test-compile</phase>
                        <goals> <goal>testCompile</goal> </goals>
                    </execution>
                </executions>
            </plugin>

```

> > 1. Can you please also check what I am doing wrong.
> 
> Sure I will check it, once I have some time 

Thanks!!!
 
> > 1. Finally, regarding testing and CI I also wanted to ask if we are going to copy the samples and *.json entity files from the parent project to this repo and set up Travis to execute them all or can we re-use them in some way?
> 
> Sure this is always in my list of things to do. We can do that.

Would you mind if I start adding CI tests within this PR mainly for maven and then I guess we can also enable the gradle tests as per #63 as well?

**Comment by sendilkumarn** [comment]
> You want me to create a PR even though it is unfinished?

I guess that is a good way to collaborate. I don't think there is any problem with that.

> From what I understand needles are mainly used to append and not to remove,

Yeah, you are correct. This is the problem. Since we are not generating the XML with the code it is difficult to do this stuff. We can have a local pom.xml copy until we find a better solution.

**Comment by pvliss** [comment]
> > You want me to create a PR even though it is unfinished?
> 
> I guess that is a good way to collaborate. I don't think there is any problem with that.
 
I think you misunderstood me. I will submit the PR is a minute. I thought you could check the commit by using the patch since the build will now be failing unless you also apply changes of the latest PR https://github.com/jhipster/generator-jhipster/issues/8991. @jdubois merged and then reverted as he found a bug while testing from the command line. Anyways the latest PR https://github.com/jhipster/generator-jhipster/pull/8999 fixes that issue as well.

> > From what I understand needles are mainly used to append and not to remove,
> 
> Yeah, you are correct. This is the problem. Since we are not generating the XML with the code it is difficult to do this stuff. We can have a local pom.xml copy until we find a better solution.

Yes, I wanted to avoid that but we can temporarily add it I guess. Will create a new ticket to the parent project for that too

**Comment by pvliss** [comment]
Sorry for the delay. I was trying to rebase and  squash the commits and it took me some time...

**Comment by sendilkumarn** [comment]
> Sorry for the delay. I was trying to rebase and squash the commits and it took me some time...

No worries. I will have look at this in next week ( starting my vacation :) )

**Comment by pvliss** [comment]
Cool. Thanks. Have a nice time :)

**Comment by pvliss** [comment]
Hello @sendilkumarn. Did you have any chance to look into this yet?

BTW, the issue with 'pom.xml' is now fixed by commit 692a962.

I re-created the TODO list here and added some more items to keep track of what needs to be done (maybe we should move it to the description?):

- [ ] There are still a few unit tests failing
- [ ] Have not tested with gradle
- [ ] Have not tested with the other samples.
- [ ] We need https://github.com/jhipster/generator-jhipster/pull/8999 accepted and merged in the parent JHipster project in order for everything to work properly.
- [x] We need to override the settings of the maven-compiler-plugin to disable annotation processing which should be only performed by kapt.
- [ ] Do not copy files from parent entity-server generator that are not needed by Kotlin, e.g  the scala Gatling test class, in order to avoid maintenance cost.
- [ ] Try to re-use as much as possible the samples from the parent JHipster project
- [ ] Update to JHipster 5.7.3 and make necessary changes to also include https://github.com/jhipster/generator-jhipster/issues/8647

**Comment by pvliss** [comment]
@sendilkumarn I am quite certain that I have found the problem and now everything works as expected. I will wait for the CI build to confirm first :pray: and then post the details and update the todo list

**Comment by pvliss** [comment]
@sendilkumarn Yes, at last everything builds properly!!! :tada: 

So the problem was with the `jackson-module-kotlin`. Removing it and enabling the parameter names for the Java compiler did the trick as Spring Boot's `org.springframework.boot.autoconfigure.jackson.JacksonAutoConfiguration` already configures the `com.fasterxml.jackson.module.paramnames.ParameterNamesModule` and everything plays out nicely. Not sure what the problem is with `jackson-module-kotlin` but maybe I will check at a later point and report back to them.

Another change was to configure the `all-open` plugin for JPA entities in order to make Hibernate happy although this was not the root of the problem.

**Comment by pvliss** [comment]
Hello @sendilkumarn. Any chance you have reviewed this yet or maybe you are waiting for SB 2.1 to be merge onto JHipster master? Also do you have any plans of what we could do next?

**Comment by amatosg** [comment]
@sendilkumarn I think @pvliss should be a member...

**Comment by pvliss** [comment]
Thanks @amatosg for the support although I am not sure if I qualify as a member. The important thing here is to move the project forward. Having said that, I realize that its OSS and it takes some time and this PR is really big and will probably need quite a large amount of @sendilkumarn's spare time. Nevertheless, I think we should do a list of what is already don, plan the next steps and start working on some of these in parallel. For example, here's an example list of high-level tasks:
- Gradually migrate/translate all Java code to Kotlin to get the full Kotlin experience
- Consider writing more idiomatic Kotlin code or use DSLs where available
- Consider using Kotlin specific libraries where it makes sense
- Support multi-platform projects which would also mean to generate the client using Kotlin as well in conjunction with Vue or React, etc.  and share code between server and client
- Perhaps also consider Kotlin native in order to generate native mobile clients

**Comment by pvliss** [comment]
@sendilkumarn As JHipster 5.8.0 has been released would you like me to continue keeping this PR aligned with the JHipster master branch?

**Comment by sendilkumarn** [comment]
@pvliss we can concentrate on the JHipster master branch.

Our next goal is to support JHipster 6, I do generally agree with all those points that you have mentioned. Feel free to open separate issues for them, we will do them gradually ❤️

**Comment by sendilkumarn** [comment]
Closing this issue as the PR is merged.

**Comment by pvliss** [comment]
> @pvliss we can concentrate on the JHipster master branch.
> 
> Our next goal is to support JHipster 6, I do generally agree with all those points that you have mentioned. Feel free to open separate issues for them, we will do them gradually

Cool. I will start by opening an issue for upgrading to 5.8.1 that was just released.

So after that, I think the next big thing is to migrate the whole server generator until we reach feature parity with the main java generator.  I am thinking of opening a master issue with sub-tasks e.g. user management, configuration,client, gateway,etc. and add "small" PRs for each one in order to make it easier to review/merge 

Hopefully the JH6/sb2.1 branch will be merged within the next couple of days and will update to that as well.
